File: content/single-module-application/src/main/java/${project_rootPath}/screen/login/LoginScreen.java
package ${project_rootPackage}.screen.login;

import io.jmix.core.CoreProperties;
import io.jmix.core.Messages;
import io.jmix.securityui.authentication.AuthDetails;
import io.jmix.securityui.authentication.LoginScreenAuthenticationSupport;
import io.jmix.ui.Notifications;
import io.jmix.ui.action.Action;
import io.jmix.ui.component.CheckBox;
import io.jmix.ui.component.ComboBox;
import io.jmix.ui.component.PasswordField;
import io.jmix.ui.component.TextField;
import io.jmix.ui.navigation.Route;
import io.jmix.ui.screen.*;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.BadCredentialsException;

import java.util.Locale;

@UiController("${project_idPrefix}_LoginScreen")
@UiDescriptor("login-screen.xml")
@Route(path = "login", root = true)
public class LoginScreen extends Screen {

    @Autowired
    private TextField<String> usernameField;

    @Autowired
    private PasswordField passwordField;

    @Autowired
    private CheckBox rememberMeCheckBox;

    @Autowired
    private ComboBox<Locale> localesField;

    @Autowired
    private Notifications notifications;

    @Autowired
    private Messages messages;

    @Autowired
    private LoginScreenAuthenticationSupport authenticationSupport;

    @Autowired
    private CoreProperties coreProperties;

    @Subscribe
    private void onInit(InitEvent event) {
        usernameField.focus();
        initLocalesField();
        initDefaultCredentials();
    }

    private void initLocalesField() {
        localesField.setOptionsMap(coreProperties.getAvailableLocales());
        localesField.setValue(coreProperties.getAvailableLocales().values().iterator().next());
    }

    private void initDefaultCredentials() {
        usernameField.setValue("admin");
        passwordField.setValue("admin");
    }

    @Subscribe("submit")
    private void onSubmitActionPerformed(Action.ActionPerformedEvent event) {
        login();
    }

    private void login() {
        String username = usernameField.getValue();
        String password = passwordField.getValue() != null ? passwordField.getValue() : "";

        if (StringUtils.isEmpty(username) || StringUtils.isEmpty(password)) {
            notifications.create(Notifications.NotificationType.WARNING)
                    .withCaption(messages.getMessage(getClass(), "emptyUsernameOrPassword"))
                    .show();
            return;
        }

        try {
            authenticationSupport.authenticate(
                    AuthDetails.of(username, password)
                            .withLocale(localesField.getValue())
                            .withRememberMe(rememberMeCheckBox.isChecked()), this);
        } catch (BadCredentialsException e) {
            notifications.create(Notifications.NotificationType.ERROR)
                    .withCaption(messages.getMessage(getClass(), "loginFailed"))
                    .withDescription(e.getMessage())
                    .show();
        }
    }
}

File: content/project/single-module-application-kotlin/src/main/kotlin/${project_rootPath}/screen/login/LoginScreen.kt
package ${project_rootPackage}.screen.login

import io.jmix.core.MessageTools
import io.jmix.core.Messages
import io.jmix.securityui.authentication.AuthDetails
import io.jmix.securityui.authentication.LoginScreenSupport
import io.jmix.ui.JmixApp
import io.jmix.ui.Notifications
import io.jmix.ui.action.Action.ActionPerformedEvent
import io.jmix.ui.component.*
import io.jmix.ui.navigation.Route
import io.jmix.ui.screen.*
import io.jmix.ui.security.UiLoginProperties
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.authentication.DisabledException
import org.springframework.security.authentication.LockedException
import java.util.*

@UiController("${normalizedPrefix_underscore}LoginScreen")
@UiDescriptor("login-screen.xml")
@Route(path = "login", root = true)
open class LoginScreen : Screen() {

    @Autowired
    private lateinit var usernameField: TextField<String>

    @Autowired
    private lateinit var passwordField: PasswordField

    @Autowired
    private lateinit var rememberMeCheckBox: CheckBox

    @Autowired
    private lateinit var localesField: ComboBox<Locale>

    @Autowired
    private lateinit var notifications: Notifications

    @Autowired
    private lateinit var messages: Messages

    @Autowired
    private lateinit var messageTools: MessageTools

    @Autowired
    private lateinit var loginScreenSupport: LoginScreenSupport

    @Autowired
    private lateinit var loginProperties: UiLoginProperties

    @Autowired
    private lateinit var app: JmixApp

    private val log = LoggerFactory.getLogger(LoginScreen::class.java)

    @Subscribe
    private fun onInit(event: InitEvent) {
        usernameField.focus()
        initLocalesField()
        initDefaultCredentials()
    }

    private fun initLocalesField() {
        localesField.apply {
            setOptionsMap(messageTools.availableLocalesMap)
            value = app.locale
            addValueChangeListener(this@LoginScreen::onLocalesFieldValueChangeEvent)
        }
    }

    private fun onLocalesFieldValueChangeEvent(event: HasValue.ValueChangeEvent<Locale>) {
        @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
        app.locale = event.value
        UiControllerUtils.getScreenContext(this).screens
                .create(this.javaClass, OpenMode.ROOT)
                .show()
    }

    private fun initDefaultCredentials() {
        val defaultUsername = loginProperties.defaultUsername
        if (!defaultUsername.isNullOrBlank() && "<disabled>" != defaultUsername) {
            usernameField.setValue(defaultUsername)
        } else {
            usernameField.value = ""
        }
        val defaultPassword = loginProperties.defaultPassword
        if (!defaultPassword.isNullOrBlank() && "<disabled>" != defaultPassword) {
            passwordField.value = defaultPassword
        } else {
            passwordField.value = ""
        }
    }

    private fun getMessage(key: String) =
        messages.getMessage(javaClass, key) ?: ""

    @Subscribe("submit")
    private fun onSubmitActionPerformed(event: ActionPerformedEvent) {
        login()
    }

    private fun login() {
        val username = usernameField.value
        val password = passwordField.value
        if (username.isNullOrBlank() || password.isNullOrBlank()) {
            notifications.create(Notifications.NotificationType.WARNING)
                .withCaption(getMessage("emptyUsernameOrPassword"))
                .show()
            return
        }
        try {
            loginScreenSupport.authenticate(
                AuthDetails.of(username, password)
                    .withLocale(localesField.value)
                    .withRememberMe(rememberMeCheckBox.isChecked), this
            )
        } catch (e: Exception) {
            when (e) {
                is BadCredentialsException,
                is DisabledException,
                is LockedException -> {
                    log.warn("Login failed for user '{}': {}", username, e.toString())
                    notifications.create(Notifications.NotificationType.ERROR)
                        .withCaption(getMessage("loginFailed"))
                        .withDescription(getMessage("badCredentials"))
                        .show()
                }
                else -> throw e
            }
        }
    }
}